include/camp/array.hpp

```diff
@@ -0,0 +1,324 @@
+         static_assert(N > 0, "Calling camp::array::front on an empty array is not allowed.");
+         return elements[0];
+      }
+
+      CAMP_HOST_DEVICE constexpr reference back() {
+         static_assert(N > 0, "Calling camp::array::back on an empty array is not allowed.");
+         return elements[size() - 1];
+      }
+
+      CAMP_HOST_DEVICE constexpr const_reference back() const {
+         static_assert(N > 0, "Calling camp::array::back on an empty array is not allowed.");
+         return elements[size() - 1];
+      }
+
+      CAMP_HOST_DEVICE constexpr pointer data() noexcept {
+         return elements;
+      }
+
+      CAMP_HOST_DEVICE constexpr const_pointer data() const noexcept {
+         return elements;
+      }
+
+      CAMP_HOST_DEVICE constexpr iterator begin() noexcept {
+         return elements;
+      }
+
+      CAMP_HOST_DEVICE constexpr const_iterator begin() const noexcept {
+         return elements;
+      }
+
+      CAMP_HOST_DEVICE constexpr const_iterator cbegin() const noexcept {
+         return elements;
+      }
+
+      CAMP_HOST_DEVICE constexpr iterator end() noexcept {
+         return &elements[size()];
+      }
+
+      CAMP_HOST_DEVICE constexpr const_iterator end() const noexcept {
+         return &elements[size()];
+      }
+
+      CAMP_HOST_DEVICE constexpr const_iterator cend() const noexcept {
+         return &elements[size()];
+      }
+
+      CAMP_HOST_DEVICE constexpr bool empty() const noexcept {
+         return size() == 0;
+      }
+
+      CAMP_HOST_DEVICE constexpr size_type size() const noexcept {
+         return N;
+      }
+
+      CAMP_HOST_DEVICE constexpr size_type max_size() const noexcept {
+         return size();
+      }
+
+      CAMP_HOST_DEVICE constexpr void fill(const T& value) {
+         for (std::size_t i = 0; i < size(); ++i) {
+            elements[i] = value;
+         }
+      }
+
+      value_type elements[N];
+   };
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator==(const array<T, N>& lhs,
+                                              const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] != rhs[i]) {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator!=(const array<T, N>& lhs,
+                                              const array<T, N>& rhs) {
+      return !(lhs == rhs);
+   }
+
+   template <class T, std::size_t N>
+   CAMP_HOST_DEVICE constexpr bool operator<(const array<T, N>& lhs,
+                                             const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] < rhs[i]) {
+            return true;
+         }
+         else if (lhs[i] == rhs[i]) {
+            continue;
+         }
+         else {
+            return false;
+         }
+      }
+
+      return false;
```

Comment:
```suggestion
      for (std::size_t i = 0; i < N; ++i) {
         if (lhs[i] >= rhs[i]) {
            return false;
         }
      }

      return true;
```
Using operator== as an example, shouldn't this operator looks more like this? 
And same thing for the other operators <=, >, >=

Reply:
I think it depends on the ordering. I might be missing something. Should {1, 3} > {1, 2} return true or false?

Reply:
It returns true. See the code example I posted in a separate comment.

Reply:
I've modified the implementation to match std::lexicographical_compare: https://en.cppreference.com/w/cpp/algorithm/lexicographical_compare

Reply:
hmmm yeah, I guess that should return true... I wasn't aware (until now) that it was a lexicographical compare and not just a raw element-wise comparison.

Reply:
@adayton1 I think you have everything correct. I should have verified before posting my comments. Sorry for the churn.

Reply:
@long58 I also learned the same lesson. 😄